apps/dashboard/pkg/migration/schemaversion/datasource_utils.go

```diff
@@ -0,0 +1,110 @@
+package schemaversion
+
+// Shared utility functions for datasource migrations across different schema versions.
+// These functions handle the common logic for migrating datasource references from
+// string names/UIDs to structured reference objects with uid, type, and apiVersion.
+
+// GetDataSourceRef creates a datasource reference object with uid, type and optional apiVersion
+func GetDataSourceRef(ds *DataSourceInfo) map[string]interface{} {
+	if ds == nil {
+		return nil
+	}
+	ref := map[string]interface{}{
+		"uid":  ds.UID,
+		"type": ds.Type,
+	}
+	if ds.APIVersion != "" {
+		ref["apiVersion"] = ds.APIVersion
+	}
+	return ref
+}
+
+// GetDefaultDSInstanceSettings returns the default datasource if one exists
+func GetDefaultDSInstanceSettings(datasources []DataSourceInfo) *DataSourceInfo {
+	for _, ds := range datasources {
+		if ds.Default {
+			return &DataSourceInfo{
+				UID:        ds.UID,
+				Type:       ds.Type,
+				Name:       ds.Name,
+				APIVersion: ds.APIVersion,
+			}
+		}
+	}
+	return nil
+}
+
+// GetInstanceSettings looks up a datasource by name or uid reference
+func GetInstanceSettings(nameOrRef interface{}, datasources []DataSourceInfo) *DataSourceInfo {
+	if nameOrRef == nil || nameOrRef == "default" {
+		return GetDefaultDSInstanceSettings(datasources)
+	}
+
+	// Check if it's a reference object without UID - should return default
+	if ref, ok := nameOrRef.(map[string]interface{}); ok {
+		if _, hasUID := ref["uid"]; !hasUID {
+			return GetDefaultDSInstanceSettings(datasources)
+		}
+		// It's a reference object with UID, search for matching UID
+		for _, ds := range datasources {
+			if uid, hasUID := ref["uid"]; hasUID && uid == ds.UID {
+				return &DataSourceInfo{
+					UID:        ds.UID,
+					Type:       ds.Type,
+					Name:       ds.Name,
+					APIVersion: ds.APIVersion,
+				}
+			}
+		}
+		return nil
```

Comment:
Do we want to return the default datasource if not found as well?

Reply:
Good catch! I just fixed it. This will make it consistent with the frontend